### ISSUE TITLE
Graphs in Maintenance GUI

### ISSUE THREAD

**Issue by robintw** [debrief/pepys-import#796]
As part of Phase 6 of the Maintenance GUI, we are adding the ability to show graphs of data in a new graphs panel.

This issue is a place for @IanMayo (and probably @clenoble) to provide information on what sort of graphs to display, plus mockups of how the graphs might look etc.

@IanMayo As a starting point, you mentioned some sort of histogram alongside the preview table - is that right? Would that be showing how many times that object is referenced? Eg. for Platform Types, the histogram would show the number of platforms that reference each PlatformType entry? And for Platforms, would it show how many sensors reference that Platform (Sensors being the main 'dependent object' for a Platform), or how many States/Contacts/etc reference that Platform (this probably makes more sense from the user perspective).

**Comment by IanMayo** [comment]
Hello @robintw - currently reviewing Rel 5 Part B. 

I imagined the workflow being:
- user selects data-type of interest
- user applies filters to reduce number of rows
- user chooses fields appropriate to what to display:
    - single column of foreign keys (nationality for a set of platfoms), "View histogram" would be offered
    - single column of numeric data - "View line plot" would be offered
    - two or more columns of numeric data, one of which is time - "View time chart" would be offered
    - two columns of numeric data, neither is time - "View scatter plot" would be offered
    - single column of location data - "View map plot"
    
Aah, there could be complexities for if we're displaying data from multiple platforms.  Aaah, ok. User selects columns of speed, time, sensor from "States" table.  Graph is scatter plot, with one series for each sensor.

**Comment by robintw** [comment]
This has been reduced in priority now, so I'm making a few notes on the investigations I did ready for any future work on this.

The key issue is finding an appropriate ASCII plotting library. There are various options, and none of them do exactly what we want.

The key features we need are:

- Ability to specify the width and height of the plot, so that it can be sized to fit inside the plotting pane (and then resized if we want to 'pop out' the pane into a popup dialog or similar) - a lot of plotting libraries don't support this
- Ability to plot a range of graphs including bar charts (horizontal histograms, but with categories not numerical data)
- Ability to use datetime data on the X axis
- Ideally good labelling of axes etc

Probably the most suitable library overall is [plotext](https://github.com/piccolomo/plotext). It allows the specification of the width and height, and seems fairly mature, but does not support horizontal bar charts or use of datetime data on the X axis. The author is fairly responsive on Github (I reported [a bug](https://github.com/piccolomo/plotext/issues/15) and got it fixed very quickly), and he says he is [working on horizontal bar chart support](https://github.com/piccolomo/plotext/issues/16)). There is no built-in way to get the ASCII text of the plot as a string, but this is easy to add - and I have already mentioned to the author that I might do a PR to add that.

An alternative is [termplotlib](https://github.com/nschloe/termplotlib) which also seems fairly mature. It supports horizontal bar charts, but also doesn't seem to support datetime data. Plotting line or scatter plots requires [gnuplot](http://www.gnuplot.info/) which would be another dependency to add which isn't so easily installable, and would potentially be complex on Windows.

So, in summary: we need a bit more than either of these libraries provide. We could add the required functionality ourselves (they're all licensed in such a way that we could), or potentially commission the library authors to add them - but they're definitely a good start how they are at the moment.

**Comment by IanMayo** [comment]
Potentially defer this, now that Pepys-Timeline has introduced browser-based interactions, we may elect to provide a more capable browser based graphical display capability.
